operator/cnp_event.go

```diff
@@ -91,11 +132,50 @@ func enableCNPWatcher() error {
 	controller.NewManager().UpdateController("cnp-to-groups",
 		controller.ControllerParams{
 			DoFunc: func(ctx context.Context) error {
-				groups.UpdateCNPInformation()
+				// Retrieves all the CNP that has currently a
+				// derivative policy and creates the new
+				// derivatives policies with the latest
+				// information from providers.  To avoid issues
+				// with rate-limiting this function will
+				// execute the addDerivative function with a
+				// max number of concurrent calls, defined on
+				// maxConcurrentUpdates.
+				cnpToUpdate := cnpCache.GetAllCNP()
+				sem := make(chan bool, maxConcurrentUpdates)
+				for _, cnp := range cnpToUpdate {
+					sem <- true
+					go func(cnp *cilium_v2.CiliumNetworkPolicy) {
+						defer func() { <-sem }()
+						addDerivativeCNP(cnp)
+					}(cnp)
+				}
 				return nil
```

Comment:
this controller returns without making sure that all go routines started on line 147 have stopped

Reply:
These goroutines will run for as long as operator main function runs, won't they? Or are you concerned with controller running for more than `RunInterval` and creating goroutines second time? I think this is a valid concern, will add a wait before return.

Reply:
Agreed, will add waitgroup here.